src/home/homework.c

```diff
@@ -1612,21 +1624,37 @@ static int home_update(UserRecord *h, Hashmap *blobs, uint64_t call_flags, UserR
         assert(h);
         assert(ret);
 
-        password_cache_load_keyring(h, &cache);
+        if (!FLAGS_SET(call_flags, SD_HOMED_UPDATE_OFFLINE)) {
+                password_cache_load_keyring(h, &cache);
 
-        r = user_record_authenticate(h, h, &cache, /* strict_verify= */ true);
-        if (r < 0)
-                return r;
-        assert(r > 0); /* Insist that a password was verified */
+                r = user_record_authenticate(h, h, &cache, /* strict_verify= */ true);
+                if (r < 0)
+                        return r;
+                assert(r > 0); /* Insist that a password was verified */
 
-        r = home_validate_update(h, &setup, &flags);
-        if (r < 0)
-                return r;
+                r = home_validate_update(h, &setup, &flags);
+                if (r < 0)
+                        return r;
+        } else {
+                /* In offline mode we skip all authentication, since we're
+                 * not propagating anything into the home area. The new home
+                 * records's authentication will still be checked when the user
+                 * next logs in, so this is fine */
+
+                r = home_basic_validate_update(h);
+                if (r < 0)
+                        return r;
+        }
 
         r = home_apply_new_blob_dir(h, blobs);
         if (r < 0)
                 return r;
 
+        if (FLAGS_SET(call_flags, SD_HOMED_UPDATE_OFFLINE)) {
```

Comment:
i am still not a fan of handing down the flags param like this. so what about this instead: just change the "verb" parameter we pass to homework if we want an offline update?

i.e. instead of passing the flags param down switch between the "update" and "update-offline" verb we pass to homework. Internally in homework you could then pass this as a boolean flag then.

all i am concerned about here is that the flags stuff is kinda specific to one verb but you pass it down for all. and different verbs might need different flags eventually, hence it appears a bit dirty to me to stick them all in a single generic field that is reused among all.

Reply:
Well flags are supposed to be extensible and composable. One idea for future flags is to support an update mode that simply amends the blob dirs instead of completely replacing them (would be useful for something like gnome-control-center to simply replace the profile picture of a user, instead of going through a whole complicated process to preserve the existing contents). That flag should be allowed to coexist w/ the OFFLINE flag.

Reply:
i am still not convinced of plumbing the flags field through all the way from dbus to the backend. i'd be more comfortable with passing through relevant fields as separate env vars, but the whole flags field not so sure, and in particular not that the backend receives it as a single field the same way for all verbs, even though we might want to support only some flags on some verbs